lib/utilities/retrieveData.js

```diff
@@ -47,24 +46,27 @@ export default function retrieveData(locations, retrieveDataFn, response, log) {
                 if (!responseDestroyed) {
                     _destroyResponse();
                 }
-                readable.emit('close');
+                readable.destroy();
```

Comment:
so this is the magic line freeing up the memory? It basically pushes null into the readable stream.  https://github.com/nodejs/node/blob/master/lib/_stream_readable.js#L189

Reply:
Yes, it is : by destroying the readable, it causes the underlying socket to be closed, so the kernel frees the bytes that have been buffered in the receive queue of the node process, and signals the other end  that it will not read from it again, causing the send buffer associated to nginx to be freed as well, and nginx getting a notification that the socket has been closed.
As for the implementation of readable.destroy, I honestly don't know. We have to be careful though : you're pointing to the branch master of node, but we're using a v4.8.4 (which is so old that it's even not available in the tags for the node project ; oldest is 4.8.6).
It just seems more reasonable to call destroy() than to emit a close event... and it closes the socket, which is what we were missing so far.

Reply:
I remember trying this trick where I push `null` into the readable stream, it didn't seem to have much effect.
I am emitting `close` here so that I can have an listener in sproxydclient to destroy to the client request in-progress. https://github.com/scality/sproxydclient/blob/rel/6.4/lib/sproxyd.js#L231
This is critical and should not be removed, otherwise the listener is never called. That listener (per docs) should drop the unconsumed data in the streams and then destroy the socket. The event name emitted can be anything, I chose `close` as I thought that it seemed appropriate. Also, having the listener is sproxydclient makes sense so that any other projects using the lib doesn't have to re-implement the fix, all they have to do is emit `close`.

Reply:
So may be the fix is to let it this way here emitting a `close` and change sproxydclient so that it does a `destroy` of the socket (it currently calls `abort`) ? Sounds better ?

Reply:
`abort` actually destroys the socket as well. https://github.com/nodejs/node/blob/master/lib/_http_client.js#L265